minidump-processor/src/processor.rs

```diff
@@ -63,6 +64,212 @@ pub struct ProcessorOptions<'a> {
+    pub fn new(subscriptions: PendingProcessorStatSubscriptions) -> Self {
+        Self {
+            subscriptions,
+            stats: Default::default(),
+        }
+    }
+
+    /// Gets (processed_thread_count, total_thread_count).
+    ///
+    /// This will panic if you didn't subscribe to
+    /// [`PendingProcessorStatSubscriptions::thread_count`].
+    pub fn get_thread_count(&self) -> (u64, u64) {
+        assert!(
+            self.subscriptions.thread_count,
+            "tried to get thread count stats, but wasn't subscribed!"
+        );
+        let stats = self.stats.lock().unwrap();
+        (stats.num_threads_processed, stats.total_threads)
+    }
+
+    /// Get count of walked frames.
+    ///
+    /// This will panic if you didn't subscribe to
+    /// [`PendingProcessorStatSubscriptions::frame_count`].
+    pub fn get_frame_count(&self) -> u64 {
+        assert!(
+            self.subscriptions.frame_count,
+            "tried to get frame count stats, but wasn't subscribed!"
+        );
+        let stats = self.stats.lock().unwrap();
+        stats.num_frames_processed
+    }
+
+    /// Get all the new walked frames since this method was last called.
+    ///
+    /// This operates via callback to allow implementation flexibility.
+    ///
+    /// This will panic if you didn't subscribe to
+    /// [`PendingProcessorStatSubscriptions::live_frames`].
+    pub fn drain_new_frames(&self, mut callback: impl FnMut(WalkedFrame)) {
+        assert!(
+            self.subscriptions.live_frames,
+            "tried to get new frames, but wasn't subscribed!"
+        );
+        let mut stats = self.stats.lock().unwrap();
+        for frame in stats.new_walked_frames.drain(..) {
+            callback(frame);
+        }
+    }
+
+    /// Get the unwalked [`ProcessState`], if it has been computed.
+    ///
+    /// This will yield `Some` exactly once.
+    ///
+    /// This will panic if you didn't subscribe to
+    /// [`PendingProcessorStatSubscriptions::unwalked_result`].
+    pub fn take_unwalked_result(&self) -> Option<ProcessState> {
+        assert!(
+            self.subscriptions.unwalked_result,
+            "tried to get unwalked result, but wasn't subscribed!"
+        );
+        let mut stats = self.stats.lock().unwrap();
+        stats.unwalked_result.take()
+    }
+
+    /// Record how many threads there are in total.
+    pub(crate) fn set_total_threads(&self, total_threads: u64) {
+        // Only bother doing this if the user cares
+        if self.subscriptions.thread_count {
+            let mut stats = self.stats.lock().unwrap();
+            stats.total_threads = total_threads;
+        }
+    }
+
+    /// Record that a thread has been processed.
+    pub(crate) fn inc_processed_threads(&self) {
+        // Only bother doing this if the user cares
+        if self.subscriptions.thread_count {
+            let mut stats = self.stats.lock().unwrap();
```

Comment:
since you worry about all these locks, did you consider making this entire reporting infrastructure a channel?  Something like to subscribe you'd have to pass in the appropriate sender to ProcessOptions and then all of the modify functions here are sending things over events, maybe including the current check on whether the subscriber cares about the event.

Reply:
The "pull" lock approach seems a lot easier to work with for the actual intended users (interactive UIs with specific refresh rates) than a "push" approach which would require the ui to keep waking up and updating state when it doesn't necessarily care about updating things.

But I am not a Concurrency Person so someone with stronger convictions could convince me otherwise.